backend/src/main/java/com/bakdata/conquery/sql/conversion/context/selects/MergedSelects.java

```diff
@@ -31,16 +32,17 @@ public class MergedSelects implements Selects {
 	List<Field<Object>> mergedSelects;
 
 	public MergedSelects(List<QueryStep> querySteps) {
-		this.primaryColumn = this.extractPrimaryColumnSelect(querySteps);
+		this.primaryColumn = this.coalescePrimaryColumns(querySteps);
 		this.validityDate = this.extractValidityDate(querySteps);
 		this.mergedSelects = this.mergeSelects(querySteps);
 	}
 
-	private Field<Object> extractPrimaryColumnSelect(List<QueryStep> querySteps) {
-		// as we join all QuerySteps / CTEs onto the same primary column,
-		// it's sufficient to obtain it from the first one in the list
-		QueryStep firstQueryStep = querySteps.iterator().next();
-		return this.mapFieldToQualifier(firstQueryStep.getCteName(), firstQueryStep.getSelects().getPrimaryColumn());
+	private Field<Object> coalescePrimaryColumns(List<QueryStep> querySteps) {
+		List<Field<Object>> primaryColumns = querySteps.stream()
+													   .map(queryStep -> this.mapFieldToQualifier(queryStep.getCteName(), queryStep.getSelects().getPrimaryColumn()))
+													   .toList();
+		return DSL.coalesce((Object) primaryColumns.get(0), primaryColumns.subList(1, primaryColumns.size()).toArray())
```

Comment:
warum musst du hier drauf coalescen?

Reply:
Wenn ich z.B. zwei CTEs verodere, kann es sein, dass ich entweder 
1. in beiden CTEs ein Element mit derselben ID habe oder
2. nur in einer der beiden CTEs ein Element ohne entsprechend übereinstimmende ID in der anderen CTE.

In jedem Fall will ich im SELECT statement für die Veroderung/Verundung aber nur einmal die ID eines jeden Elements auswählen. Da ich bei der Veroderung nicht einfach die ID der ersten CTE nehmen kann (weil nullable), coalesce ich auf die ID.

Reply:
**TL;DR: Ich könnte das `coalesce()` nur im Falle einer OR-Verknüpfung nutzen.**

Die `MergedSelects` bilden ja die Selects der zu joinenden Tabellen ab. Für die ID muss ich dann z.B. entweder `cte_1.id` oder `cte_2.id` selektieren und joine dann beide CTEs auf deren jeweilige ID.

Bezogen auf das Beispiel unten: Im Falle von einem AND hätte ich tatsächlich in jedem Fall nur IDs die sowohl in `concept_0_vs` und `concept_1_vs` vorkommen, weil sonst das inner join nicht erfüllt ist. Da könnte ich mir das `coalesce()` sparen.
Bei einem OR joine ich auch auf die gemeinsamen IDs, habe aber zusätzlich noch alle Einträge die in `concept_0_vs`, aber nicht in `concept_1_vs` vorkommen und alle Einträge, die in `concept_1_vs` aber nicht in `concept_0_vs` vorkommen. 

Wenn ich also statt dem `coalesce("concept_0_vs"."pid", "concept_1_vs"."pid")` z.B. nur `"concept_0_vs"."pid"` selektiere, dann habe ich bei einem OR bzw. full outer join ggf. Einträge, die nur in `concept_1_vs` vorhanden sind und demnach bei `"concept_0_vs"."pid"` `null` ergeben.

```sql
# AND (does not need coalesce)
"concept_0_vs_OR_concept_1_vs" as (
    select "concept_0_vs"."pid" as "primary_column"
    from "concept_0_vs"
      inner join "concept_1_vs"
        on "concept_0_vs"."pid" = "concept_1_vs"."pid"
  )

# OR (needs coalesce)
"concept_0_vs_OR_concept_1_vs" as (
    select coalesce("concept_0_vs"."pid", "concept_1_vs"."pid") as "primary_column"
    from "concept_0_vs"
      full outer join "concept_1_vs"
        on "concept_0_vs"."pid" = "concept_1_vs"."pid"
  )
```

Reply:
> TL;DR: Ich könnte das coalesce() nur im Falle einer OR-Verknüpfung nutzen.


ahhh, verstehe ich jetzt. Danke für die Ausführliche Erklärung :)